pkg/resiliency/resiliency.go

```diff
@@ -1,132 +1,504 @@
+			continue
+		}
+	}
+	return r
+}
+
+// New creates a new Resiliency.
+func New() *Resiliency {
+	return &Resiliency{
+		timeouts:        make(map[string]time.Duration),
+		retries:         make(map[string]*retry.Config),
+		circuitBreakers: make(map[string]*breaker.CircuitBreaker),
+		actorCBCaches:   make(map[string]*lru.Cache),
+		serviceCBs:      make(map[string]*circuitBreakerInstances),
+		componentCBs: &circuitBreakerInstances{
+			cbs: make(map[string]*breaker.CircuitBreaker, 10),
+		},
+		services:   make(map[string]PolicyNames),
+		actors:     make(map[string]ActorPolicyNames),
+		components: make(map[string]PolicyNames),
+		routes:     make(map[string]PolicyNames),
+	}
+}
+
+// DecodeConfiguration reads in a single resiliency configuration.
+func (r *Resiliency) DecodeConfiguration(c *resiliency_v1alpha.Resiliency) error {
+	if c == nil {
+		return nil
+	}
+	var err error
+	policies := c.Spec.Policies
+	buildingBlocks := c.Spec.BuildingBlocks
+
+	for name, t := range policies.Timeouts {
+		if r.timeouts[name], err = parseDuration(t); err != nil {
+			return fmt.Errorf("invalid duration %q, %s: %w", name, t, err)
+		}
+	}
+
+	for name, t := range policies.Retries {
+		rc := retry.DefaultConfig()
+		m, err := toMap(t)
+		if err != nil {
+			return err
+		}
+		if err = retry.DecodeConfig(&rc, m); err != nil {
+			return fmt.Errorf("invalid retry configuration %q: %w", name, err)
+		}
+		r.retries[name] = &rc
+	}
+
+	for name, t := range policies.CircuitBreakers {
+		var cb breaker.CircuitBreaker
+		m, err := toMap(t)
+		if err != nil {
+			return err
+		}
+		if err = config.Decode(m, &cb); err != nil {
+			return fmt.Errorf("invalid retry configuration %q: %w", name, err)
+		}
+		cb.Name = name
+		cb.Initialize()
+		r.circuitBreakers[name] = &cb
+	}
+
+	for name, t := range buildingBlocks.Services {
+		r.services[name] = PolicyNames{
+			Timeout:        t.Timeout,
+			Retry:          t.Retry,
+			CircuitBreaker: t.CircuitBreaker,
+		}
+		r.serviceCBs[name] = &circuitBreakerInstances{
+			cbs: make(map[string]*breaker.CircuitBreaker, 10),
+		}
+	}
+
+	for name, t := range buildingBlocks.Actors {
+		scope, err := ParseActorCircuitBreakerScope(t.CircuitBreakerScope)
+		if err != nil {
+			return err
+		}
+		r.actors[name] = ActorPolicyNames{
+			PolicyNames: PolicyNames{
+				Timeout:        t.Timeout,
+				Retry:          t.Retry,
+				CircuitBreaker: t.CircuitBreaker,
+			},
+			CircuitBreakerScope: scope,
+		}
+		r.actorCBCaches[name], err = lru.New(t.CircuitBreakerCacheSize)
+		if err != nil {
+			return err
+		}
 	}
 
-	for _, body := range resp.GetResiliencies() {
-		resiliences = appendResiliency(resiliences, body, log)
+	for name, t := range buildingBlocks.Components {
+		r.components[name] = PolicyNames{
+			Timeout:        t.Timeout,
+			Retry:          t.Retry,
+			CircuitBreaker: t.CircuitBreaker,
+		}
+	}
+
+	for name, t := range buildingBlocks.Routes {
+		name = strings.TrimPrefix(name, "/") // Ignore slash prefix.
```

Comment:
Does this handle multiple "/"?

Reply:
It doesn't need to. It's based on the path which throughout Dapr can have 1 or 0 prefix forward slashes.

Reply:
If it's just an index into the map, is there a reason to ignore it? Seems like it'll work as a key regardless of the slash.

Reply:
Good question. We have examples that both use prefix slashes and not. I actually had to add this line in order to normalize the values for comparison.